### ISSUE TITLE
snapcraft has more strict linting unexpectedly 

### ISSUE THREAD

**Issue by dilyn-corner** [canonical/snapcraft#5017]
### Bug Description

When building a snap using at least revision 12523 to build snaps which built successfully with revisions like 12503 (latest/edge versus 8.x/edge), snapcraft will now emit some error messages.

### To Reproduce

Install snapcraft from latest/edge, build the snap with `snapcraft`.

### Environment

Ubuntu 23.10, building with `snapcraft --use-lxd`.

### snapcraft.yaml

```yaml
name: my-snap-name # you probably want to 'snapcraft register <name>'
base: core24 # the base snap is the execution environment for this snap
version: '0.1' # just for humans, typically '1.2+git' or '1.3.2'
summary: Single-line elevator pitch for your amazing snap # 79 char long summary
description: |
  This is my-snap's description. You have a paragraph or two to tell the
  most important story about your snap. Keep it under 100 words though,
  we live in tweetspace and your description wants to look good in the snap
  store.

grade: devel # must be 'stable' to release into candidate/stable channels
confinement: devmode # use 'strict' once you have the right plugs and slots

parts:
  my-part:
    # See 'snapcraft plugins'
    plugin: nil

apps:
  my-app:
    command: bash
    environment:
      foo: 1
```

### Relevant log output

```shell
dilyn@Ares:~/test -> snapcraft 
Starting snapcraft, version 8.3.2.post71+git3508cf39
Logging execution to '/home/dilyn/.local/state/snapcraft/log/snapcraft-20240905-191836.634265.log'
Bad snapcraft.yaml content:
- input should be a valid string (in field 'apps.my-app.environment.foo')
```

### Additional context

Quoting `1` results in successful builds (sans an error message unrelated to this issue).

**Comment by lengau** [comment]
I think this has roughly the same cause as https://github.com/canonical/snapcraft/issues/4978

**Comment by dilyn-corner** [comment]
Makes sense -- all the revisions I have where this is working are before the pydantic 2 shift.

We can close as duplicate.

**Comment by mr-cal** [comment]
I got the [`craft-parts` fix](https://github.com/canonical/craft-parts/pull/816) into snapcraft 8.4.0 but will need to get the [`craft-application` fix](https://github.com/canonical/craft-application/pull/450) in 8.4.1.

**Comment by mr-cal** [comment]
This issue actually occurs in craft-grammar: https://github.com/canonical/craft-grammar/issues/49

**Comment by mr-cal** [comment]
Fixed via #5051
